### DISCUSSION TITLE
How to create lists longer than 1000 elements?

### DISCUSSION THREAD

**Discussion by J-Cake** [nushell/nushell#9505] [Q&A]
Hi everyone, I'm trying to parse a large amount of data using NuShell, which is working wonderful, but I'm having to reduce my dataset in order to get NuShell to play along. Am I doing something wrong or is there a hard limit on the length of lists. The lists seem to count to 999 and begin a new list, at this point the sorting of the list is reset.

My code is the following: It is desinged to sort a number of base64-encoded JSON records placed in a file ending with .JSON, separated by newlines. Each entry contains the key 'Company Name' or a case-insensitive version of it. 

```nu
def parse_json [] {
  let out = ($in | from json)
  if ($out | describe) == 'string' { $out | parse_json } else { $out }
}

 (ls *.json) | each { |i| $i | get name } | each { open --raw | split row -r '[\s]+' } | flatten | filter { |i| not ($i | is-empty) } | each { decode base64 | parse_json } | sort-by 'Company Name'
```

**Comment by fdncred** [comment]
Nushell will create separate lists/tables at 1000 rows or at a given timeout interval. The way you avoid that is to use the `collect` command. You can read more about it with `help collect`.

I thought of this one too `1..1005 | str join (char nl)`.

If you have problems, ping me here with some sample data, and I'll try an help more.

**Reply by fdncred** [reply]
The more I think about it, the `collect` is for joining tables/lists less than 1000 entries. I'm not sure there is a way to join a table/list of 1005 entries, but you can use `str join (char nl)` to make a contiguous list of text (not a structured nushell list).

**Reply by J-Cake** [reply]
Hm, that could work, but I was hoping keep it in tabular form. See I'm dealing with thousands of records, which I need to sort and organise, is this limit hard-coded or does it exist in a config file somewhere, where I can adjust it?

**Edit**: Alternatively, if you can suggest to me a way of sorting data in a streaming fashion, that could work too

**Reply by fdncred** [reply]
I don't think you can sort data in a streaming fashion at all. In order to sort things, I think you have to have them all. So, you could do something like:
```
let biglist = (some pipeline)
let sorted_list = ($biglist | sort)
$sorted_list
```
The data will still be sorted but when nushell displays it, it'll be in chunks of 1000

**Comment by ghost** [comment]
You can display it properly using `$your_data | explore`. (for me it then displays a single list without any interruptions).  
If you want to export it to a file i would recommend using `| save result.json` if a maschiene needs it or `| to md --pretty | save -r a.md` if its destined for humans.

In case you have performance issues (or it beeing slow) i would like to recommend you
[dataframes](https://www.nushell.sh/book/dataframes.html), which exist specifically to deal with big datasets.  
To use them you have to use a version of nushell which was compiled with support for them (`cargo build --release --features default,dataframe`)

**Reply by fdncred** [reply]
we recently changed features. to get all commands you have to use `--featrues=dataframe,extra` and to get extra speed I use `--features=dataframe,extra,mimalloc`

**Reply by J-Cake** [reply]
@Jan9103 Thanks, I did end up discovering `explore`, which provided me with the functionality I needed to look at the data. 

As for sorting it, I ended up having to write a separate two-pass program which inserts the sort-keys into a BTree, as well as the offset of where to find the record in the database, and iterate that. 

I was hoping to be able to do this directly in Nu, but I ended up turning to the tRUSTy ol' crab

**Comment by fire** [comment]
I would probably solve this as a a microbatching where for a 1000 entries:

1. the result is fetched in 1000 entry chunks
2. the 1000 entries are chunked
3. the result is stored in object storage (like s3) or a database like cockroachdb / sqlite
